lib/HLSL/DxilValidation.cpp

```diff
@@ -3988,6 +3988,114 @@ static void ValidateMetadata(ValidationContext &ValCtx) {
   }
 
   const hlsl::ShaderModel *SM = ValCtx.DxilMod.GetShaderModel();
+  // validate that any wavesize tags don't appear outside their expected shader
+  // models. Validate only 1 tag exists per entry point.
+  // Don't do this validation if the shader is non-compute
+  if (SM->IsCS() || SM->IsLib()) {
+
+    NamedMDNode *TA = pModule->getNamedMetadata("dx.entryPoints");
+    if (TA != nullptr) {
+      for (unsigned i = 0, end = TA->getNumOperands(); i < end; ++i) {
+        MDTuple *TANode = dyn_cast<MDTuple>(TA->getOperand(i));
+        if (TANode->getNumOperands() < 5) {
+          ValCtx.EmitMetaError(TANode, ValidationRule::MetaWellFormed);
+          return;
+        }
+        // get access to the digit that represents the metadata number that
+        // would store entry properties
+        const llvm::MDOperand &mOp =
+            TANode->getOperand(TANode->getNumOperands() - 1);
+        // the final operand to the entry points tuple should be a constant int.
+        if (mOp == nullptr ||
+            (mOp.get())->getMetadataID() != Metadata::MDTupleKind) {
+          continue;
+        }
+
+        // get access to the node that stores entry properties
+        MDTuple *TANode2 =
+            dyn_cast<MDTuple>(TANode->getOperand(TANode->getNumOperands() - 1));
```

Comment:
This is retrieving the same value as `mOp` above, but casting it to MDTuple. Is that right? 

Again I'm unsure what "TA" refers to. This should be the entry properties tuple and naming it accordingly would make the code easier to follow.